internal/pkg/agent/application/upgrade/step_unpack_test.go

```diff
@@ -0,0 +1,192 @@
+	REGULAR fileType = iota
+	DIRECTORY
+	SYMLINK
+)
+
+type files struct {
+	fType   fileType
+	path    string
+	content string
+	mode    fs.FileMode
+}
+
+func (f files) Name() string {
+	return path.Base(f.path)
+}
+
+func (f files) Size() int64 {
+	return int64(len(f.content))
+}
+
+func (f files) Mode() fs.FileMode {
+	return f.mode
+}
+
+func (f files) ModTime() time.Time {
+	return time.Unix(0, 0)
+}
+
+func (f files) IsDir() bool {
+	return f.fType == DIRECTORY
+}
+
+func (f files) Sys() any {
+	return nil
+}
+
+type createArchiveFunc func(*testing.T, []files) (string, error)
+
+func TestUpgrader_unpack(t *testing.T) {
+	type args struct {
+		version          string
+		archiveGenerator createArchiveFunc
+		archiveFiles     []files
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr assert.ErrorAssertionFunc
+	}{
+		{
+			name: "targz with file before containing folder",
+			args: args{
+				version: "1.2.3",
+				archiveFiles: []files{
+					{fType: DIRECTORY, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64", mode: fs.ModeDir | (fs.ModePerm & 0o750)},
+					{fType: REGULAR, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/" + agentCommitFile, content: "abcdefghijklmnopqrstuvwxyz", mode: fs.ModePerm & 0o640},
+					{fType: REGULAR, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef/package.version", content: "1.2.3", mode: fs.ModePerm & 0o640},
+					{fType: DIRECTORY, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data", mode: fs.ModeDir | (fs.ModePerm & 0o750)},
+					{fType: DIRECTORY, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef", mode: fs.ModeDir | (fs.ModePerm & 0o750)},
+					{fType: REGULAR, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef/" + agentName, content: "Placeholder for the elastic-agent binary", mode: fs.ModePerm & 0o750},
+					{fType: DIRECTORY, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef/components", mode: fs.ModeDir | (fs.ModePerm & 0o750)},
+					{fType: REGULAR, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef/components/comp1", content: "Placeholder for component", mode: fs.ModePerm & 0o750},
+					{fType: REGULAR, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/data/elastic-agent-abcdef/components/comp1.spec.yml", content: foo_component_spec, mode: fs.ModePerm & 0o640},
+					{fType: SYMLINK, path: "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64/" + agentName, content: "data/elastic-agent-abcdef/" + agentName, mode: fs.ModeSymlink | (fs.ModePerm & 0o750)},
+				},
+				archiveGenerator: func(t *testing.T, i []files) (string, error) {
+					return createTarArchive(t, "elastic-agent-1.2.3-SNAPSHOT-linux-x86_64.tar.gz", i)
+				},
+			},
+			want:    "abcdef",
+			wantErr: assert.NoError,
+		},
+	}
+	for _, tt := range tests {
+		prevTop := paths.Top()
+		t.Run(tt.name, func(t *testing.T) {
+			testTop := t.TempDir()
+			paths.SetTop(testTop)
```

Comment:
There was a race condition that @faec just fixed related to adjusting paths in tests. Would it be possible to just pass in the top directory into the unpack function. That way the test can use that directory versus changing a global.

Reply:
Removed the references to `paths.Data()` that required us to set `paths.Top()` in the unit test.